### ISSUE TITLE
[MIR] Implement debug info

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#31005]
Currently we do not emit any debuginfo from the MIR translator.

It doesn’t seem like there should be many issues implementing support, other than `DebugLoc` wanting a `ast::NodeId`, which eventually gets passed down to `trans::debuginfo::metadata::scope_metadata`.

Precondition to this work is to either decouple `trans::debuginfo` and its DIScope-building-related functionality from the `NodeId` or create a independent code for MIR’s use.

**Comment by nagisa** [comment]
It might certainly be possible that by the translation time we do not retain enough information to build good debuginfo. In that case MIR itself should be adjusted to carry enough data.

**Comment by eddyb** [comment]
While working on #32803 I've found that we're still missing information on arguments, closure captures and returns.

For arguments, I tried leaving them anonymous, as gcc and clang do for `void foo(A, B, C) {}` in C++.
However, gdb doesn't display the value at all, while lldb makes an effort and uses `(null)` as the name.

Our options involve:
* the name of single-binding patterns (and anonymous for the rest)
* a pretty-printed pattern in `ArgDecl`
* a source fragment extracted from the pattern span, in `ArgDecl`
* some mapping from `Name` to `Lvalue` (could be translated to DWARF deref/offset expressions)

That last option would also help with closure captures, as they are effectively fields of the first argument.

As for returns, old trans seems to use the closing brace of the function (`span.hi`?), now that I think about it.
We're probably not going to be able to do anything interesting for each early `return` as they're just a branch and will remain that way for the foreseeable future.
